app/services/courses/validate_custom_age_range_service.rb

```diff
@@ -0,0 +1,37 @@
+module Courses
+  class ValidateCustomAgeRangeService
+    def execute(age_range_in_years, course)
+      error_message = "#{age_range_in_years} is invalid. You must enter a valid age range."
+      valid_regex_pattern = Regexp.new(/^(?<from>\d{1,2})_to_(?<to>\d{1,2})$/)
+
+      if valid_regex_pattern.match(age_range_in_years)
+        from_age = get_ages(age_range_in_years, valid_regex_pattern)["from"]
+        to_age = get_ages(age_range_in_years, valid_regex_pattern)["to"]
+      end
+
+      if !valid_regex_pattern.match(age_range_in_years)
```

Comment:
Running the regex match twice feels like the wrong thing to do here. I think it should either be:

```ruby
      age_range_matches = valid_regex_pattern.match(age_range_in_years)

      if age_range_matches
        from_age = get_ages(age_range_in_years, valid_regex_pattern)["from"]
        to_age = get_ages(age_range_in_years, valid_regex_pattern)["to"]
      end

      # Handle errors separately
      if !age_range_matches
        ...
      end
```

or, because I still find handling the `!age_range_matches` this way, without an `else`, a little consfusing:

```ruby
      if valid_regex_pattern.match(age_range_in_years)
        from_age = get_ages(age_range_in_years, valid_regex_pattern)["from"]
        to_age = get_ages(age_range_in_years, valid_regex_pattern)["to"]

        if from_age_invalid?(from_age)
          course.errors.add(:age_range_in_years, error_message)
        elsif to_age_invalid?(to_age)
          course.errors.add(:age_range_in_years, error_message)
        elsif to_age - from_age < 4
          course.errors.add(:age_range_in_years, "#{age_range_in_years} is invalid. Your age range must cover at least 4 years.")
        end
      else
        course.errors.add(:age_range_in_years, error_message) 
      end
```

Reply:
Yea that's way clearer. Thanks for that.